### DISCUSSION TITLE
Internal error "refs/remotes/main/HEAD is not a symbolic ref" in log buffer and unrecognised glob patterns

### DISCUSSION THREAD

**Discussion by mscfd** [magit/magit#4938] [Support] [answered]
Trying to understand glob pattern issue (see below) I used magit-toggle-git-debug to get some debug information. This ended up that for a number of log options for the magit log buffer ("current", "HEAD", "local branches" etc.) I get

> 128 git … symbolic-ref refs/remotes/main/HEAD
> fatal: ref refs/remotes/main/HEAD is not a symbolic ref

in the magit process buffer. As far as I can tell this is due to "HEAD" in argument tuples for magit-log-setup-buffer, like in 
``(magit-log-setup-buffer (list "HEAD" "--branches"))``
occuring frequently in magit-log.el.
It is obviously nothing fatal, but somehow the internal git log from magit does not seem to handle this properly.

The reason why I came upon this: I tried to setup my own custom function using
```
(defun magit-log-my-stuff (&optional args files)
  (interactive (magit-log-arguments))
  "Show log for my branches."
  (magit-log-setup-buffer
   (list "--branches" "--remotes=origin/mystuff/[xy]*")
   args files))
```
which works fine so far (using transient to add it to the log options). However, for more complex glob patterns, which work on the command line, like ``git log --remotes={origin/mystuff/[xy]*,other/bar/*}``, magit fails to show anything at all, using ``(list "--branches" "--remotes={origin/mystuff/[xy]*,other/bar/*}")`` in the my-stuff function above. Probably same with "--branches". This looks more like a feature request than an actual bug, as magit probably does its own parsing and does not yet understand shell glob patterns completely? According to output in the magit process buffer, git log is not invoked to obtain the log graph.

**Comment by kyleam** [comment, marked as answer]
>I used magit-toggle-git-debug to get some debug information. This ended up that for a number of log options for the magit log buffer ("current", "HEAD", "local branches" etc.) I get
>
> > 128 git … symbolic-ref refs/remotes/main/HEAD
> > fatal: ref refs/remotes/main/HEAD is not a symbolic ref

This is unlikely to indicate a problem, for the reason described in the [manual section](https://magit.vc/manual/magit/Debugging-Tools.html) on `magit-toggle-git-debug`:

    Also note that just because git exits with a non-zero exit status
    and prints an error message that usually doesn’t mean that it is
    an error as far as Magit is concerned, which is another reason we
    usually hide these error messages. Whether some error message is
    relevant in the context of some unexpected behavior has to be
    judged on a case by case basis.

I haven't traced the code, but I would bet it's coming from `magit-log-wash-rev`s call to `magit-format-ref-labels`:

https://github.com/magit/magit/blob/152478b01f37e3aee1369e815985f184d8c864c5/lisp/magit-git.el#L2384-L2386

> However, for more complex glob patterns, which work on the command line, like `git log --remotes={origin/mystuff/[xy]*,other/bar/*}`,

Are you getting confused by your shell doing [brace expansion](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Brace-Expansion)?  In bash, for example, `git` itself would up seeing

    '--remotes=origin/mystuff/[xy]*' '--remotes=other/bar/*'

if it was passed the arguments spelled in your example.

You can do `export GIT_TRACE=1` in your shell to verify.

Those expanded arguments are what would need to be passed as arguments to the `magit-log-setup-buffer` call because neither Magit nor the `git` call underneath will expand those braces for you.

> magit probably does its own parsing and does not yet understand shell glob patterns completely?

Magit passes those arguments through to `git log`.  Instrument and step through `magit--insert-log` to convince yourself.

**Comment by kyleam** [comment]
(I'm going to convert this to a discussion because I don't see evidence of a bug, but we can of course create a distilled issue if there ends up being a bug here.)

**Comment by mscfd** [comment]
Thanks a lot for the detailed reply and sorry for the false alarm. I was (and still am) confused that the git log call does not show up in the magit process buffer.
Providing (list ... "--remotes=pattern1" "--remotes=pattern2") to magit-log-setup-buffer gives the expected result.
